test/integration/health_check_integration_test.cc

```diff
@@ -334,5 +335,161 @@ TEST_P(TcpHealthCheckIntegrationTest, SingleEndpointTimeoutTcp) {
   EXPECT_EQ(1, test_server_->counter("cluster.cluster_1.health_check.failure")->value());
 }
 
+class GrpcHealthCheckIntegrationTest : public testing::TestWithParam<Network::Address::IpVersion>,
+                                       public HealthCheckIntegrationTestBase {
+public:
+  GrpcHealthCheckIntegrationTest() : HealthCheckIntegrationTestBase(GetParam()) {}
+
+  void TearDown() override {
+    cleanupHostConnections();
+    cleanUpXdsConnection();
+  }
+
+  // Adds a gRPC active health check specifier to the given cluster, and waits for the first health
+  // check probe to be received.
+  void initGrpcHealthCheck(uint32_t cluster_idx) {
+    auto& cluster_data = clusters_[cluster_idx];
+    auto health_check = addHealthCheck(cluster_data.cluster_);
+    health_check->mutable_grpc_health_check();
+
+    // Introduce the cluster using compareDiscoveryRequest / sendDiscoveryResponse.
+    EXPECT_TRUE(compareDiscoveryRequest(Config::TypeUrl::get().Cluster, "", {}, {}, {}, true));
+    sendDiscoveryResponse<envoy::config::cluster::v3::Cluster>(
+        Config::TypeUrl::get().Cluster, {cluster_data.cluster_}, {cluster_data.cluster_}, {}, "55");
+
+    // Wait for upstream to receive HC request.
+    grpc::health::v1::HealthCheckRequest request;
+    ASSERT_TRUE(cluster_data.host_upstream_->waitForHttpConnection(
+        *dispatcher_, cluster_data.host_fake_connection_));
+    ASSERT_TRUE(cluster_data.host_fake_connection_->waitForNewStream(*dispatcher_,
+                                                                     cluster_data.host_stream_));
+    ASSERT_TRUE(cluster_data.host_stream_->waitForGrpcMessage(*dispatcher_, request));
+    ASSERT_TRUE(cluster_data.host_stream_->waitForEndStream(*dispatcher_));
+
+    EXPECT_EQ(cluster_data.host_stream_->headers().getPathValue(), "/grpc.health.v1.Health/Check");
+    EXPECT_EQ(cluster_data.host_stream_->headers().getContentTypeValue(),
+              Http::Headers::get().ContentTypeValues.Grpc);
+    EXPECT_EQ(cluster_data.host_stream_->headers().getHostValue(), cluster_data.name_);
+  }
+
+  // Send a gRPC message with the given headers and health check response message.
+  void sendGrpcResponse(uint32_t cluster_idx,
+                        const Http::TestResponseHeaderMapImpl& response_headers,
+                        const grpc::health::v1::HealthCheckResponse& health_check_response) {
+    clusters_[cluster_idx].host_stream_->encodeHeaders(response_headers, false);
+    clusters_[cluster_idx].host_stream_->sendGrpcMessage(health_check_response);
+    clusters_[cluster_idx].host_stream_->finishGrpcStream(Grpc::Status::WellKnownGrpcStatus::Ok);
```

Comment:
would testing `grpc-status` ever be relevant? I think if there's a non-200 status response, this gets used instead, but most cases besides 200/serving will anyway go to failure.

Reply:
I don't think the tests require anything other than a 200-status to be sent (as you say it's just a failure), but if there is a need then this can easily be changed.